File: app/looper.py
import _thread
import copy
import threading
from strip_ctl import Strip
from time import sleep
from datetime import datetime

DELAY = 0.0250

strips = [
    Strip(1, "esp32c6-00.lan", 67),
    Strip(2, "esp32c6-01.lan", 83),
]

modes = [strip.modes["NightRider"] for strip in strips]

# Debug mode control
_debug_mode = False
_loop_control = {
    "state": "running",  # "running" or "paused"
    "iterations_remaining": -1,  # -1 means unlimited
    "event": threading.Event(),
}
_loop_control["event"].set()  # Start in running state


def set_debug_mode(enabled):
    global _debug_mode
    _debug_mode = enabled


def get_debug_mode():
    return _debug_mode


def loop_control(iterations=None, next_state=None):
    """Control the loop for testing purposes. Only works in debug mode.

    If iterations is provided, run that many iterations then auto-pause.
    If only next_state is provided, apply it immediately.
    """
    if not _debug_mode:
        return {"error": "debug mode not enabled"}

    if iterations is not None:
        # Run N iterations then pause (next_state is ignored when stepping)
        _loop_control["iterations_remaining"] = iterations
        _loop_control["state"] = "running"
        _loop_control["event"].set()
    elif next_state == "pause":
        _loop_control["state"] = "paused"
        _loop_control["event"].clear()
    elif next_state == "running":
        _loop_control["state"] = "running"
        _loop_control["iterations_remaining"] = -1
        _loop_control["event"].set()

    return {
        "state": _loop_control["state"],
        "iterations_remaining": _loop_control["iterations_remaining"],
    }


def get_loop_state():
    """Get current loop state."""
    return {
        "state": _loop_control["state"],
        "iterations_remaining": _loop_control["iterations_remaining"],
        "debug_mode": _debug_mode,
    }


def loop():
    global _loop_control
    while True:
        # Debug mode: wait if paused
        if _debug_mode:
            _loop_control["event"].wait()

        a = datetime.now()
        for mode in modes:
            mode.update()
        b = datetime.now()
        c = (b - a).microseconds / 1000000.0
        if DELAY > c:
            sleep(DELAY - c)

        # Debug mode: handle iteration counting
        if _debug_mode and _loop_control["iterations_remaining"] > 0:
            _loop_control["iterations_remaining"] -= 1
            if _loop_control["iterations_remaining"] == 0:
                _loop_control["state"] = "paused"
                _loop_control["event"].clear()


def set_mode(m):
    global modes
    modes = [strip.modes[m] for strip in strips]
    for mode in modes:
        if hasattr(mode, "load_cb"):
            mode.load_cb({"set_delay": set_delay})


def get_current_mode():
    return modes[0].name


def get_modes():
    return strips[0].modes


def get_opts():
    return modes[0].get_opts()


def set_opts(val):
    for mode in modes:
        mode.set_opts(copy.deepcopy(val))


def get_delay():
    return DELAY


def set_delay(num):
    global DELAY
    DELAY = num


def get_brightness():
    return strips[0].get_brightness()


def set_brightness(num):
    for strip in strips:
        strip.set_brightness(num)


def get_strips():
    """Return strip configuration for API consumers."""
    return [
        {
            "id": strip.dev_id,
            "hostname": strip.UDP_IP,
            "port": strip.UDP_PORT,
            "num_leds": strip.NUM_LEDS,
        }
        for strip in strips
    ]


def configure_strip(strip_id, hostname=None, port=None):
    """Configure a strip's network settings (debug mode only).

    This allows tests to redirect UDP packets to localhost listeners.
    """
    if not _debug_mode:
        return {"error": "debug mode not enabled"}

    for strip in strips:
        if strip.dev_id == strip_id:
            if hostname is not None:
                strip.UDP_IP = hostname
            if port is not None:
                strip.UDP_PORT = port
            return {
                "id": strip.dev_id,
                "hostname": strip.UDP_IP,
                "port": strip.UDP_PORT,
            }

    return {"error": f"strip {strip_id} not found"}


def start_loop():
    _thread.start_new_thread(loop, ())

File: app_rs/src/looper.rs
use crate::modes::{create_mode, Mode};
use crate::opts::OptValue;
use crate::strip::Strip;
use serde::Serialize;
use std::collections::HashMap;
use std::sync::{Arc, Mutex, Condvar};
use std::thread;
use std::time::{Duration, Instant};

pub struct LooperState {
    pub strips: Vec<Strip>,
    pub modes: Vec<Box<dyn Mode + Send>>,
    pub delay: f64,
    pub debug_mode: bool,
    pub loop_state: String,  // "running" or "paused"
    pub iterations_remaining: i64,  // -1 means unlimited
    pub condvar: Arc<(Mutex<bool>, Condvar)>,
}

impl LooperState {
    pub fn new(debug_mode: bool) -> Self {
        let strips = vec![
            Strip::new(1, "esp32c6-00.lan", 67),
            Strip::new(2, "esp32c6-01.lan", 83),
        ];
        
        let mut modes: Vec<Box<dyn Mode + Send>> = Vec::new();
        for strip in &strips {
            modes.push(create_mode("NightRider", strip));
        }
        
        Self {
            strips,
            modes,
            delay: 0.025,
            debug_mode,
            loop_state: "running".to_string(),
            iterations_remaining: -1,
            condvar: Arc::new((Mutex::new(true), Condvar::new())),
        }
    }
    
    pub fn get_current_mode_name(&self) -> String {
        if let Some(mode) = self.modes.first() {
            mode.name().to_string()
        } else {
            "Unknown".to_string()
        }
    }
    
    pub fn set_mode(&mut self, mode_name: &str) {
        let mut new_modes: Vec<Box<dyn Mode + Send>> = Vec::new();
        for strip in &self.strips {
            let mut mode = create_mode(mode_name, strip);
            mode.load_cb(&|_d: f64| { 
                // We can't modify self.delay from here directly
                // Store the request and handle it after
            });
            new_modes.push(mode);
        }
        self.modes = new_modes;
        
        // Apply delay changes from load_cb - since load_cb doesn't actually do anything
        // in this simplified version, we'll just set known delays for specific modes
        match mode_name {
            "Solid" | "White" | "Off" => self.delay = 0.250,
            _ => self.delay = 0.025,
        }
    }
    
    pub fn get_brightness(&self) -> u8 {
        self.strips.first().map(|s| s.get_brightness()).unwrap_or(255)
    }
    
    pub fn set_brightness(&mut self, val: u8) {
        for strip in &mut self.strips {
            strip.set_brightness(val);
        }
    }
    
    pub fn get_opts(&self) -> HashMap<String, OptValue> {
        if let Some(mode) = self.modes.first() {
            mode.get_opts()
        } else {
            HashMap::new()
        }
    }
    
    pub fn set_opts(&mut self, opts: HashMap<String, OptValue>) {
        for mode in &mut self.modes {
            mode.set_opts(opts.clone());
        }
    }
    
    pub fn get_strips(&self) -> Vec<StripInfo> {
        self.strips.iter().map(|s| StripInfo {
            id: s.dev_id,
            hostname: s.udp_ip.clone(),
            port: s.udp_port,
            num_leds: s.num_leds,
        }).collect()
    }
    
    pub fn configure_strip(&mut self, strip_id: u8, hostname: Option<String>, port: Option<u16>) -> Result<StripInfo, String> {
        if !self.debug_mode {
            return Err("debug mode not enabled".to_string());
        }
        
        for strip in &mut self.strips {
            if strip.dev_id == strip_id {
                if let Some(h) = hostname {
                    strip.udp_ip = h;
                }
                if let Some(p) = port {
                    strip.udp_port = p;
                }
                return Ok(StripInfo {
                    id: strip.dev_id,
                    hostname: strip.udp_ip.clone(),
                    port: strip.udp_port,
                    num_leds: strip.num_leds,
                });
            }
        }
        
        Err(format!("strip {} not found", strip_id))
    }
    
    pub fn loop_control(&mut self, iterations: Option<i64>, next_state: Option<String>) -> LoopControlResult {
        if !self.debug_mode {
            return LoopControlResult {
                state: self.loop_state.clone(),
                iterations_remaining: self.iterations_remaining,
                error: Some("debug mode not enabled".to_string()),
            };
        }
        
        if let Some(iters) = iterations {
            // Run N iterations then pause
            self.iterations_remaining = iters;
            self.loop_state = "running".to_string();
            let (lock, cvar) = &*self.condvar;
            let mut running = lock.lock().unwrap();
            *running = true;
            cvar.notify_all();
        } else if let Some(state) = next_state {
            match state.as_str() {
                "pause" => {
                    self.loop_state = "paused".to_string();
                    let (lock, _) = &*self.condvar;
                    let mut running = lock.lock().unwrap();
                    *running = false;
                }
                "running" => {
                    self.loop_state = "running".to_string();
                    self.iterations_remaining = -1;
                    let (lock, cvar) = &*self.condvar;
                    let mut running = lock.lock().unwrap();
                    *running = true;
                    cvar.notify_all();
                }
                _ => {}
            }
        }
        
        LoopControlResult {
            state: self.loop_state.clone(),
            iterations_remaining: self.iterations_remaining,
            error: None,
        }
    }
    
    pub fn get_loop_state(&self) -> LoopStateResponse {
        LoopStateResponse {
            state: self.loop_state.clone(),
            iterations_remaining: self.iterations_remaining,
            debug_mode: self.debug_mode,
        }
    }
}

#[derive(Debug, Clone, Serialize)]
pub struct StripInfo {
    pub id: u8,
    pub hostname: String,
    pub port: u16,
    pub num_leds: usize,
}

#[derive(Debug, Clone, Serialize)]
pub struct LoopControlResult {
    pub state: String,
    pub iterations_remaining: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct LoopStateResponse {
    pub state: String,
    pub iterations_remaining: i64,
    pub debug_mode: bool,
}

pub struct Looper {
    state: Arc<Mutex<LooperState>>,
}

impl Looper {
    pub fn new(debug_mode: bool) -> Self {
        let state = Arc::new(Mutex::new(LooperState::new(debug_mode)));
        
        // Clone state for the loop thread
        let state_clone = Arc::clone(&state);
        
        thread::spawn(move || {
            Self::loop_thread(state_clone);
        });
        
        Self { state }
    }
    
    fn loop_thread(state: Arc<Mutex<LooperState>>) {
        loop {
            let start = Instant::now();
            
            // Check if we should wait (debug mode + paused)
            {
                let state_guard = state.lock().unwrap();
                if state_guard.debug_mode && state_guard.loop_state == "paused" {
                    // Wait on condition variable
                    let condvar = Arc::clone(&state_guard.condvar);
                    drop(state_guard);
                    
                    let (lock, cvar) = &*condvar;
                    let mut running = lock.lock().unwrap();
                    while !*running {
                        running = cvar.wait(running).unwrap();
                    }
                }
            }
            
            // Update modes and send
            {
                let mut state_guard = state.lock().unwrap();
                let len = state_guard.strips.len().min(state_guard.modes.len());
                
                for i in 0..len {
                    // We need to update each mode with its corresponding strip
                    // Use unsafe to get mutable references to different elements
                    let strip = &mut state_guard.strips[i] as *mut Strip;
                    let mode = &mut state_guard.modes[i];
                    unsafe {
                        mode.update(&mut *strip);
                    }
                }
                
                // Handle iteration counting in debug mode
                if state_guard.debug_mode && state_guard.iterations_remaining > 0 {
                    state_guard.iterations_remaining -= 1;
                    if state_guard.iterations_remaining == 0 {
                        state_guard.loop_state = "paused".to_string();
                        let (lock, _) = &*state_guard.condvar;
                        let mut running = lock.lock().unwrap();
                        *running = false;
                    }
                }
            }
            
            // Calculate sleep time
            let elapsed = start.elapsed();
            let delay = {
                let state_guard = state.lock().unwrap();
                Duration::from_secs_f64(state_guard.delay)
            };
            
            if elapsed < delay {
                thread::sleep(delay - elapsed);
            }
        }
    }
    
    pub fn get_state(&self) -> Arc<Mutex<LooperState>> {
        Arc::clone(&self.state)
    }
}
